gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/converter/EnvelopeSchemaConverter.java

```diff
@@ -1,152 +1,145 @@
-      GenericRecord outputRecord = deserializePayload(payload, payloadSchema);
-      if (this.fieldRemover.isPresent()) {
-        payloadSchema = this.fieldRemover.get().removeFields(payloadSchema);
-      }
-      return new SingleRecordIterable<>(AvroUtils.convertRecordSchema(outputRecord, payloadSchema));
-    } catch (IOException | SchemaRegistryException | ExecutionException e) {
-      throw new DataConversionException(e);
+  protected byte[] getPayloadBytes(GenericRecord inputRecord) throws Exception {
+    Optional<Object> payloadValue = AvroUtils.getFieldValue(inputRecord, payloadField);
+    if (!payloadValue.isPresent()) {
+      throw new Exception("Payload with key " + payloadField + " not found in the record");
     }
-  }
 
-  /**
-   * Get payload field from GenericRecord and convert to byte array
-   */
-  public byte[] getPayload(GenericRecord inputRecord, String payloadFieldName) {
-    ByteBuffer bb = (ByteBuffer) inputRecord.get(payloadFieldName);
-    byte[] payloadBytes;
+    ByteBuffer bb = (ByteBuffer) payloadValue.get();
     if (bb.hasArray()) {
-      payloadBytes = bb.array();
+      return bb.array();
     } else {
-      payloadBytes = new byte[bb.remaining()];
+      byte[] payloadBytes = new byte[bb.remaining()];
       bb.get(payloadBytes);
+      String hexString = new String(payloadBytes, StandardCharsets.UTF_8);
+      return DatatypeConverter.parseHexBinary(hexString);
     }
-    String hexString = new String(payloadBytes, StandardCharsets.UTF_8);
-    return DatatypeConverter.parseHexBinary(hexString);
+  }
+
+  protected Schema fetchLatestPayloadSchema() throws Exception {
+    Schema latestPayloadSchema = (Schema)registry.getLatestSchemaByTopic(payloadSchemaTopic);
+    latestPayloadReader = new GenericDatumReader<>(latestPayloadSchema);
+    return latestPayloadSchema;
   }
 
   /**
-   * Deserialize payload using payload schema
+   * Convert the payload in the input record to a schema'ed object of the latest schema
+   *
+   * @param inputRecord the input record
+   * @return the schema'ed payload object
    */
-  public GenericRecord deserializePayload(byte[] payload, Schema payloadSchema) throws IOException, ExecutionException {
-    Decoder decoder = this.decoderFactory.binaryDecoder(payload, null);
-    GenericDatumReader<GenericRecord> reader = this.readers.get(payloadSchema);
-    return reader.read(null, decoder);
+  protected P upConvertPayload(GenericRecord inputRecord) throws DataConversionException {
+    try {
+      Schema payloadSchema = getPayloadSchema(inputRecord);
+      // Set writer schema
+      latestPayloadReader.setSchema(payloadSchema);
+
+      byte[] payloadBytes = getPayloadBytes(inputRecord);
+      Decoder decoder = DecoderFactory.get().binaryDecoder(payloadBytes, null);
+
+      // 'latestPayloadReader.read' will convert the record from 'payloadSchema' to the latest payload schema
+      return latestPayloadReader.read(null, decoder);
+    } catch (Exception e) {
```

Comment:
Catch the specific exceptions here if possible.

Reply:
Since we don't do specific exception handling, catching specific exceptions is effectively the same as catching a general `Exception`.